osidb/api_views.py

```diff
@@ -840,6 +840,25 @@ class AffectView(
     http_method_names = get_valid_http_methods(ModelViewSet)
     permission_classes = [IsAuthenticatedOrReadOnly]
 
+    def get_serializer(self, *args, **kwargs):
+        """
+        Updates the serializer to contain ps_component if it was missing and purl is present.
+        Only minimal required validation is done, as it is also present in the Affect model.
+        """
+        data = kwargs.get("data")
+
+        if data and data.get("purl") and not data.get("ps_component"):
+            new_data = data.copy()
+            purl = data["purl"]
+            try:
+                ps_component = PackageURL.from_string(purl).name
+            except Exception as exc:
+                raise ValidationError(f"Invalid purl '{purl}': {exc}")
+            new_data["ps_component"] = ps_component
+            kwargs["data"] = new_data
```

Comment:
I don't think this is the right place to perform such field population or performing any validations. This should be responsibility of either model layer or in this case more probably serializer level. We already perform some field prepopulation on serializer level by overriding the create and update methods there so I would use those as a reference. Also this approach still treats `ps_component` as required field on API which should no longer be true as it will be possible to have it blank in the request no.

Reply:
I agree that the view is not a good place for it. Even the serializer is not ideal and the business logic should eventually move out of there but at least for now some of it is there. The validation should ideally be in the model validations and setting the component based on the PURL could eg. be in a pre-save signal or some part of the model save procedure.

Reply:
@JakubFrejlach I tried to override `create` in `AffectSerializer`/`AffectPostSerializer`, but without success, `update` itself was okay. If you prefer to use this approach here, I need some help to understand what is wrong with current serializers. I will try to update the API to describe that at least one of `purl` and `ps_component` is required.

Reply:
@jobselko I am not sure, what that override didn't work for you but as @osoukup wrote, doing that on the model layer, you can avoid this whole hustle. Just setting the ps_component not required at the serializer should be enough for you to successfully pass through serializer and then do the rest on the model layer.